zio-kafka/src/main/scala/zio/kafka/consumer/internal/Runloop.scala

```diff
@@ -541,41 +541,43 @@ private[consumer] final class Runloop(
-          .fromQueue(commandQueue)
-          .merge(ZStream.repeatWithSchedule(Command.PeriodicPoll, Schedule.spaced(pollFrequency)))
-          .aggregateAsync(ZSink.collectAllN[Command](commandQueueSize))
-          .mapZIO { commands =>
-            def logPollStart(state: State): UIO[Unit] =
-              ZIO
-                .logTrace(
-                  s"Starting poll with ${state.pendingRequests.size} pending requests and ${state.pendingCommits.size} pending commits"
-                )
+  def run(stop: Ref[Boolean]): ZIO[Scope, Throwable, Any] = {
+    def logPollStart(state: State): UIO[Unit] =
+      ZIO
+        .logTrace(
+          s"Starting poll with ${state.pendingRequests.size} pending requests and ${state.pendingCommits.size} pending commits"
+        )
 
-            for {
-              state      <- stateRef.get
-              _          <- ZIO.logDebug(s"Processing ${commands.size} commands: ${commands.mkString(",")}")
-              isShutdown <- isShutdown
-              handleCommand = if (isShutdown) handleShutdown _ else handleOperational _
-              updatedState <- ZIO.foldLeft(commands)(state)(handleCommand)
-
-              shouldPoll =
-                updatedState.isSubscribed && (updatedState.pendingRequests.nonEmpty || updatedState.pendingCommits.nonEmpty || updatedState.assignedStreams.isEmpty)
-              updatedStateAfterPoll <- if (shouldPoll) logPollStart(updatedState) *> handlePoll(updatedState)
-                                       else ZIO.succeed(updatedState)
-              _ <- stateRef.set(updatedStateAfterPoll)
-            } yield ()
-          }
-          .takeUntilZIO(_ => stop.get)
+    def shouldPoll(state: State): Boolean =
+      state.isSubscribed && (state.pendingRequests.nonEmpty || state.pendingCommits.nonEmpty || state.assignedStreams.isEmpty)
+
+    ZStream
+      .fromQueue(commandQueue)
+      .timeoutFail[Throwable](RunloopTimeout)(runloopTimeout)
+      .merge(ZStream.repeatWithSchedule(Command.Poll, Schedule.spaced(pollFrequency)))
+      .aggregateAsync(ZSink.collectAllN[Command](commandQueueSize))
+      .takeUntilZIO(_ => stop.get)
+      .runFoldZIO(State.initial) { case (state, commands) =>
+        for {
+          _          <- ZIO.logTrace(s"Processing ${commands.size} commands: ${commands.mkString(",")}")
+          isShutdown <- isShutdown
+          handleCommand = if (isShutdown) handleShutdown _ else handleOperational _
+          updatedState <- ZIO.foldLeft(commands)(state)(handleCommand)
+
+          updatedStateAfterPoll <- if (shouldPoll(updatedState))
+                                     logPollStart(updatedState) *> handlePoll(updatedState)
+                                   else ZIO.succeed(updatedState)
+          _ <-
+            commandQueue
+              .offer(Command.Poll)
+              .when(
+                shouldPoll(updatedStateAfterPoll)
+              ) // Immediately poll again, after processing all new queued commands
```

Comment:
😆 Yes, it can be that simple!

BTW, not sure what you think about this. I always prefer comments on a line before the statement.
```suggestion
           // Immediately poll again, after processing all new queued commands
           _ <-
            commandQueue
              .offer(Command.Poll)
              .when(shouldPoll(updatedStateAfterPoll))
```